core/src/main/java/org/apache/brooklyn/core/typereg/AbstractTypePlanTransformer.java

```
@@ -0,0 +1,130 @@
+     * and performs common validation and customisation of the items created.
+     * <p>
+     * this includes:
+     * <li> setting the {@link AbstractBrooklynObjectSpec#catalogItemId(String)}
+     */
+    @Override
+    public Object create(final RegisteredType type, final RegisteredTypeLoadingContext context) {
+        try {
+            return validate(new RegisteredTypeKindVisitor<Object>() {
+                @Override protected Object visitSpec(RegisteredType type) {
+                    try { 
+                        AbstractBrooklynObjectSpec<?, ?> result = createSpec(type, context);
+                        result.catalogItemId(type.getId());
+                        return result;
+                    } catch (Exception e) { throw Exceptions.propagate(e); }
+                }
+                @Override protected Object visitBean(RegisteredType type) {
+                    try { 
+                        return createBean(type, context);
+                    } catch (Exception e) { throw Exceptions.propagate(e); }
+                }
+                
+            }.visit(type), type, context);
+        } catch (UnsupportedTypePlanException e) {
+            // no logging
+            throw Exceptions.propagate(e);
+        } catch (Exception e) {
+            Exceptions.propagateIfFatal(e);
+            log.debug("Could not instantiate "+type+" (rethrowing): "+Exceptions.collapseText(e));
+            throw Exceptions.propagate(e);
+        }
+    }
+    
+    protected <T> T validate(T createdObject, RegisteredType type, RegisteredTypeLoadingContext context) {
+        if (createdObject==null) return null;
+        // TODO validation based on the constraint, throw UnsupportedTypePlanException with details if not matched
+        return createdObject;
+    }
+
+    protected abstract AbstractBrooklynObjectSpec<?,?> createSpec(RegisteredType type, RegisteredTypeLoadingContext context) throws Exception;
```

`throws Exception` works for tests, but doesn't it mask failure points that should be handled in implementations?